pkg/client/cache/reflector.go

```diff
@@ -337,16 +304,23 @@ func (r *Reflector) ListAndWatch(stopCh <-chan struct{}) error {
 			}
 			return nil
 		}
-		if err := r.watchHandler(w, &resourceVersion, resyncCh, stopCh); err != nil {
-			if err != errorResyncRequested && err != errorStopRequested {
+
+		if err := r.watchHandler(w, &resourceVersion, stopCh); err != nil {
+			if err != errorStopRequested {
 				glog.Warningf("%s: watch of %v ended with: %v", r.name, r.expectedType, err)
 			}
 			return nil
 		}
-		if r.canForceResyncNow() {
-			glog.V(4).Infof("%s: next resync planned for %#v, forcing now", r.name, r.nextResync)
-			return nil
-		}
+		/*
```

Comment:
I believe the issue is that the watch times out and closes, but if there is a change to an object in etcd before the watch is restarted then we will miss the notification of that change.  I ran into this issue running e2es, specifically the "should fail a job" tests.  The first would always pass, and the 2nd would always fail regardless of order.  I looked at the logs and I could see the watch closing about the time the job would expire and the controller never saw the notifications.  Switching to never closing the watch handled this case.